File: backend/app/api/partners.py
from uuid import UUID
from fastapi import APIRouter, Depends, HTTPException, Response, status


from sqlalchemy.exc import IntegrityError
from sqlalchemy.orm import Session

from app.core.deps import (
    get_current_admin,
    get_current_user,
    get_db,
    get_pagination,
)
from app.models.user import User
from app.schemas.partner import (
    PartnerCreate,
    PartnerListResponse,
    PartnerPublic,
    PartnerUpdate,
)
from app.services.partner_service import (
    create_partner,
    delete_partner,
    get_partner,
    list_partners,
    update_partner,
)

router = APIRouter(prefix="/partners", tags=["partners"])


@router.get("", response_model=PartnerListResponse)
def list_partners_endpoint(
    pagination: tuple[int, int] = Depends(get_pagination),
    search: str | None = None,
    type: str | None = None,
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
):
    page, page_size = pagination
    items, total = list_partners(db, page, page_size, search, type)
    return {"items": items, "total": total, "page": page, "page_size": page_size}


@router.get("/{partner_id}", response_model=PartnerPublic)
def get_partner_endpoint(
    partner_id: UUID,
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
):
    partner = get_partner(db, partner_id)
    if not partner:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Partner not found")
    return partner


@router.post("", response_model=PartnerPublic, status_code=status.HTTP_201_CREATED)
def create_partner_endpoint(
    data: PartnerCreate,
    db: Session = Depends(get_db),
    admin: User = Depends(get_current_admin),
):
    try:
        partner = create_partner(db, data)
    except IntegrityError:
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail="Partner with this tax number already exists",
        )
    return partner


@router.put("/{partner_id}", status_code=status.HTTP_204_NO_CONTENT)
def update_partner_endpoint(
    partner_id: UUID,
    data: PartnerUpdate,
    db: Session = Depends(get_db),
    admin: User = Depends(get_current_admin),
):
    try:
        partner = update_partner(db, partner_id, data)
    except IntegrityError:
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail="Partner with this tax number already exists",
        )
    if not partner:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Partner not found")
    return Response(status_code=status.HTTP_204_NO_CONTENT)


@router.delete("/{partner_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_partner_endpoint(
    partner_id: UUID,
    db: Session = Depends(get_db),
    admin: User = Depends(get_current_admin),
):
    deleted = delete_partner(db, partner_id)
    if not deleted:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Partner not found")
    return Response(status_code=status.HTTP_204_NO_CONTENT)

File: web/src/lib/partners.ts
import { useQuery, useMutation, useQueryClient } from '@tanstack/react-query'
import api from './api'
import { Partner, PartnerCreate, PartnerUpdate } from '../types/partner'

interface ListParams {
  search: string
  type: string
  page: number
  pageSize: number
}

export const listPartners = ({ search, type, page, pageSize }: ListParams) =>
  useQuery({
    queryKey: ['partners', { search, type, page, pageSize }],
    queryFn: async () => {
      const res = await api.get('/partners', {
        params: { search, type, page, page_size: pageSize },
      })
      return res.data as {
        items: Partner[]
        total: number
        page: number
        page_size: number
      }
    },
  })

export const getPartner = (id: string) =>
  useQuery({
    queryKey: ['partners', id],
    queryFn: async () => {
      const res = await api.get(`/partners/${id}`)
      return res.data as Partner
    },
    enabled: !!id,
  })

export const createPartner = () => {
  const queryClient = useQueryClient()
  return useMutation({
    mutationFn: (payload: PartnerCreate) =>
      api.post('/partners', payload).then((res) => res.data as Partner),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['partners'] })
    },
  })
}

export const updatePartner = (id: string) => {
  const queryClient = useQueryClient()
  return useMutation({
    mutationFn: (payload: PartnerUpdate) =>
      api.put(`/partners/${id}`, payload).then((res) => res.data as Partner),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['partners'] })
      queryClient.invalidateQueries({ queryKey: ['partners', id] })
    },
  })
}

export const deletePartner = () => {
  const queryClient = useQueryClient()
  return useMutation({
    mutationFn: (id: string) => api.delete(`/partners/${id}`).then((res) => res.data),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['partners'] })
    },
  })
}

